File: 03 hard/search problems/980_unique_paths_iii.js
/**
 * @param {number[][]} grid
 * @return {number}
 */
var uniquePathsIII = function(grid) {
    let start = [-1, -1] 
    let directions = [
        [0, 1], 
        [0, -1], 
        [1, 0], 
        [-1, 0]
    ]
    let zeroes = 0

    for (var i = 0; i < grid.length; i++) {
        for (var j = 0; j < grid[i].length; j++) {
            if (grid[i][j] == 1) {
                start[0] = i 
                start[1] = j
            } else if (grid[i][j] == 0) zeroes++
        }
    }

    let paths = 0

    function getUniquePath(x, y, zero) {
        if (grid[x][y] == 2) {
           if (zero - 1 == zeroes) paths++;
            return;
        }

        if (grid[x][y] === -1 || grid[x][y] == -2) return;

        grid[x][y] = -2;

        for (var m = 0; m < directions.length; m++) {
            const i = x + directions[m][0];
            const j = y + directions[m][1];
            if (i >= 0 && i < grid.length && j >= 0 && j < grid[0].length)
                getUniquePath(i, j, zero + 1);
        }
        grid[x][y] = 0;
    }

    getUniquePath(start[0], start[1], 0)
    return paths
};

File: 03 hard/search problems/graph search/980_unique_paths_iii.py
class Solution(object):
    def uniquePathsIII(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """
        start = [0, 0]
        directions = [
            [1, 0], [0, 1], [-1, 0], [0, -1]
        ]
        squares = 0 

        rows = len(grid) 
        cols = len(grid[0])

        global paths
        paths = 0

        for i in range(rows): 
            for j in range(cols): 
                if grid[i][j] == 1: 
                    start[0] = i 
                    start[1] = j 
                if grid[i][j] == 0: 
                    squares += 1

        def dfs(row, col, squares_walked): 
            global paths
            if grid[row][col] == 2: 
                if squares_walked - 1 == squares: # minus one cuz we also count the ending square as a step
                    paths += 1
                return 
            
            if grid[row][col] == -1 or grid[row][col] == -2: 
                return 

            # marked as read
            grid[row][col] = -2

            for direction in directions: 
                i = row + direction[0]
                j = col + direction[1] 
                if 0 <= i < rows and 0 <= j < cols: 
                    dfs(i, j, squares_walked + 1) 
            
            # restore steps
            grid[row][col] = 0
        
        dfs(start[0], start[1], 0) 
        return paths 
